mantidproject/mantid
# Migrate build script to Python

**Description of work.**

    Create 1:1 port from bash build script to posix_build.py
    
    Creates a python posix build script, that is as close to 1:1 with the
    original bash script.
    
    Switching to Python allows us to long term:
    - Create a module of shared snippets (such as check CMake invocation, or
    run using XVFB ...etc.)
    - Have more developers interact with build scripts (as they are in
    Python)
    - Get to use linters to find common bugs
    - Allow us to stop throwing more and more items into the buildscript
    - Chunk down the existing buildscript and refactor it long-term

Refactoring will be in a future PR
Adding in better OS env support will be in the PR after, the idea being we can pass `RELEASE_TYPE=DEBUG` or `TIMEOUT=x` in the environment variables. Currently the former requires us to put debug in the name.

**To test:**
- Check original build script and python side-by-side. I've tried to make it as easy as possible to follow, and the only new functions are at the top and are marked internal with `_`
- Look at the build logs carefully

Part of #31069 - Getting prerequisite coverage data as a job in Jenkins required an increasing number of hacks to the build script which tipped me to doing this


*This does not require release notes* because it only affects our build systems



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
